### PR TITLE
Add zig linker support 

### PR DESCRIPTION
**Pull request by RossComputerGuy**
Adds support for Zig's linker (zld)

Merges #13493, #12293, and #12953.

Context from https://github.com/mesonbuild/meson/pull/13493#issuecomment-2269602518

> What we really need here is to have a single PR that contains all the necessary changes:
> 
>     * this patch
> 
>     * the patch from [linkers: basic support for the 'zig cc' linker #12293](https://github.com/mesonbuild/meson/pull/12293)
> 
>     * a patch to enable zig in the CI workflows, see [Add Zig to ubuntu-rolling image, CI test #12953](https://github.com/mesonbuild/meson/pull/12953)
> 
> 
> And a second PR that adds zig to the CI images _without_ adding it to the CI workflows, see #12953 again.
> 
> Currently everything is spread out quite a bit and it's difficult to get a handle on whether it all works.

### CODE REVIEW

**Comment by thesamesam** [comment]
Please don't use merge commits.

**Review by eli-schwartz — CHANGES_REQUESTED** [review]
In fact, merge commits are explicitly a violation of the Contributing docs: https://mesonbuild.com/Contributing.html#keeping-pull-requests-up-to-date

We will NOT merge any PR that contains one.

**Comment by RossComputerGuy** [comment]
> In fact, merge commits are explicitly a violation of the Contributing docs: https://mesonbuild.com/Contributing.html#keeping-pull-requests-up-to-date

Sorry, didn't know that one. Should I squash/cherry pick instead?

File: ci/ciimage/ubuntu-rolling/install.sh [hunk 1]
```diff
@@ -50,6 +51,18 @@ dub_fetch dubtestproject
 dub build dubtestproject:test1 --compiler=ldc2
 dub build dubtestproject:test2 --compiler=ldc2
 
+# zig
+
+ZIG_ARCHIVE=$(wget -qO- https://ziglang.org/download/index.json | jq -r ".master[\"$(uname -m)-linux\"].tarball")
```
Comment (thesamesam) [inline ci/ciimage/ubuntu-rolling/install.sh:56]: I'm not sure if we want to pull master for this. That'll lead to flaky CI. If we do want that, it needs justification.

Comment (RossComputerGuy) [reply ci/ciimage/ubuntu-rolling/install.sh:56]: Yeah, we can probably use the latest release. I just need to figure out the jq to read the 2nd object.

Comment (RossComputerGuy) [reply ci/ciimage/ubuntu-rolling/install.sh]: Alright, we're now using a tagged release instead of master.

**Comment by QuLogic** [comment]
Zig is in the CI image now, I think, so this just needs a rebase and dropping that last commit?

**Comment by RossComputerGuy** [comment]
> so this just needs a rebase and dropping that last commit?

Done

**Review by dcbaker — APPROVED** [review]

**Pull request merged**